Doc/extending/newtypes_tutorial.rst

```diff
@@ -0,0 +1,865 @@
+
+We want to expose our instance variables as attributes. There are a
+number of ways to do that. The simplest way is to define member definitions::
+
+   static PyMemberDef Noddy_members[] = {
+       {"first", T_OBJECT_EX, offsetof(NoddyObject, first), 0,
+        "first name"},
+       {"last", T_OBJECT_EX, offsetof(NoddyObject, last), 0,
+        "last name"},
+       {"number", T_INT, offsetof(NoddyObject, number), 0,
+        "noddy number"},
+       {NULL}  /* Sentinel */
+   };
+
+and put the definitions in the :c:member:`~PyTypeObject.tp_members` slot::
+
+   .tp_members = Noddy_members,
+
+Each member definition has a member name, type, offset, access flags and
+documentation string.  See the :ref:`Generic-Attribute-Management` section
+below for details.
+
+A disadvantage of this approach is that it doesn't provide a way to restrict the
+types of objects that can be assigned to the Python attributes.  We expect the
+first and last names to be strings, but any Python objects can be assigned.
+Further, the attributes can be deleted, setting the C pointers to *NULL*.  Even
+though we can make sure the members are initialized to non-*NULL* values, the
+members can be set to *NULL* if the attributes are deleted.
+
+We define a single method, :meth:`Noddy.name()`, that outputs the objects name as the
+concatenation of the first and last names. ::
+
+   static PyObject *
+   Noddy_name(NoddyObject *self)
+   {
+       if (self->first == NULL) {
+           PyErr_SetString(PyExc_AttributeError, "first");
+           return NULL;
+       }
+       if (self->last == NULL) {
+           PyErr_SetString(PyExc_AttributeError, "last");
+           return NULL;
+       }
+       return PyUnicode_FromFormat("%S %S", self->first, self->last);
+   }
+
+The method is implemented as a C function that takes a :class:`Noddy` (or
+:class:`Noddy` subclass) instance as the first argument.  Methods always take an
+instance as the first argument. Methods often take positional and keyword
+arguments as well, but in this case we don't take any and don't need to accept
+a positional argument tuple or keyword argument dictionary. This method is
+equivalent to the Python method::
+
+   def name(self):
+      return "%s %s" % (self.first, self.last)
+
+Note that we have to check for the possibility that our :attr:`first` and
+:attr:`last` members are *NULL*.  This is because they can be deleted, in which
+case they are set to *NULL*.  It would be better to prevent deletion of these
+attributes and to restrict the attribute values to be strings.  We'll see how to
+do that in the next section.
+
+Now that we've defined the method, we need to create an array of method
+definitions::
+
+   static PyMethodDef Noddy_methods[] = {
+       {"name", (PyCFunction) Noddy_name, METH_NOARGS,
+        "Return the name, combining the first and last name"
+       },
+       {NULL}  /* Sentinel */
+   };
+
+(note that we used the :const:`METH_NOARGS` flag to indicate that the method
```

Comment:
This hanging comment in parenthesis looks weird. What if move it up?
```rst
Now that we've defined the method, we need to create an array of method
definitions (note that we used the :const:`METH_NOARGS` flag to indicate that the method
is expecting no arguments other than *self*)::
```

Or keep it where it was in the original text?

Reply:
Hmm, they look fine to me?